### PR TITLE
Version tracking of images

### LINKED ISSUE #682: Version tracking of images
**Description**
Images can be updated by the users and the history of all the images must be stored.
The latest version of the image must be shown up. 

Something similar to the GitHub comments:
Selection_172

Image versioning, locking, and reverting must be possible.

Thanks!

### PR DESCRIPTION
**Pull request by divyabaid16**
Fixes #682  (<=== Add issue number here)

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

Find the attached screenshot:
Selection_209

[//]: # (To mark checkbox write 'x' within the square brackets)

* [x] PR is descriptively titled 📑 and links the original issue above 🔗
* [ ] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR -- or run tests locally with `rake test`
* [x] code is in uniquely-named feature branch and has no merge conflicts 📁
* [x] screenshots/GIFs are attached 📎 in case of UI updation
* [ ] ask `@publiclab/mapknitter-reviewers` for help, in a comment below

Thanks!

### CODE REVIEW

**Comment by divyabaid16** [comment]
GIF:
Peek 2019-07-17 17-16

**Comment by divyabaid16** [comment]
@jywarren @cesswairimu @gauravano

**Comment by divyabaid16** [comment]
@publiclab/mapknitter-reviewers

**Comment by cesswairimu** [comment]
Hey @divyabaid16 this is really cool!!...I have a few suggestions
- Only allow logged in users to move the images so 'anonymous' users should not be able to edit images unless the map itself is anonymous
 - Only show the last edit of a particular user e.g say user `divya` updated 2 hours ago, then 2 minutes ago just show `divya updated 2 minutes ago`
  

What do you think?

**Comment by kaustubh-nair** [comment]
Oooh this is really cool @divyabaid16  :tada:

**Comment by divyabaid16** [comment]
Thank you @cesswairimu @kaustubh-nair.
I just implemented the first part that "Only logged in users can edit the warpables unless the map itself is anonymous".
And regarding the second one, I was thinking how the UI should be, like where should a user click to see the edits of a particular author. Should we display the list of all the users who edited the map? Or where should this option be present? Maybe we can discuss this in another issue. Is it fine @cesswairimu 

Thanks!

**Comment by cesswairimu** [comment]
Awesome @divyabaid16. Yeah we can open another issue for discussion on how we could best display  this. Would you mind adding some tests?  This should be ready after that. Thanks

**Comment by divyabaid16** [comment]
Hi @cesswairimu I tried to add the test but it is failing. Actually, I'm kind of new to the testing part, so can you please provide your inputs?
Here is the error:
Selection_211
Thanks!

**Comment by cesswairimu** [comment]
Sure @divyabaid16, pulling this in a few and will take a look

**Comment by cesswairimu** [comment]
@divyabaid16 Where are you calling the `revert` method on the ui?

**Comment by divyabaid16** [comment]
I have use ajax for that. It is present at the bottom.

**Comment by jywarren** [comment]
Oh wow this is really coming along! Re: UI, maybe we should make it a small white "clock icon" button below each image?

**Comment by jywarren** [comment]
This is really impressive 😄 ✊

**Comment by divyabaid16** [comment]
Hi @jywarren I have added the clock icon. Just had a bit confusion, i.e should I remove the "Image history" button? Or keep both of them i.e, the clock icon button and the "Image history" button?

**Comment by jywarren** [comment]
Oh sorry - i guess i meant that the clock could be instead of `image history` - isn't it on a per-image basis? Or are there two buttons? Would you mind uploading a new gif? Thanks!

**Comment by divyabaid16** [comment]
@jywarren please find the updated GIF:

Peek 2019-07-20 10-42

Thanks!

File: app/views/images/_index.html.erb [hunk 1]
```diff
@@ -94,6 +101,41 @@
...
+                <%= link_to '' ,images_revert_path(id: warpable.id, version:version), class: 'fa fa-undo', title: 'Revert to this version' %>
+              </button>
+            <% end %>
+          </div>
+        </div>
+      </td>
     <% end %>
   </tbody>
 </table>
+<script>
```
Comment (cesswairimu) [inline app/views/images/_index.html.erb]: Hi @divyabaid16, I broke this script when I change the functionality of `revert`...could you update(sorry my js sucks) it please or maybe we can remove it? Thanks

**Comment by divyabaid16** [comment]
Ya, now the script is not needed. I just removed it. 
Thanks!

**Comment by cesswairimu** [comment]
Awesome thanks...I bet this is ready to go then we can do a follow to improve the design?

**Comment by divyabaid16** [comment]
Ya, that would be great:)

**Comment by divyabaid16** [comment]
Thanks to you @cesswairimu for helping me with this and working collaboratively. :balloon:

**Comment by jywarren** [comment]
This looks awesome. Just want to be really clear on the current functionality. Is this correct?

1. when you move the image, it makes a new "revision" displayed in the list. revisions are any change in coordinates in x,y of any corner
2. when you click one of the revisions, it makes a new revision cloning that old one, adding the clone to the 'top of the stack'?
3. OR it deletes all revisions after the one you clicked on?
4. anyone may make a revision

Anything else? Actually once we correct this exact documentation, perhaps we can add it to the README as well, just to make sure it's clear to people? What do you think? 

Finally, i think we should (either now or in follow-up) plan to make it really clear that clicking on a revision reverts. Maybe it should be a list, but instead of the whole item being a button, it should have a row of buttons on the left side of each revision field that says `Revert`? What do you think?

Thank you, this is really cool!

**Comment by divyabaid16** [comment]
Currently, when we click on a previous revision, it is cloned at put up on the top of stack. But this doesn't looks good. 
So I will soon be making change to the code such that clicking on the version, it just goes back to that version without deleting all the forward versions. 

So it would be like-- I click on the version, I'm reverted there. Then I click on the latest version, I'm reverted to the latest one. And it would work smoothly. 

Thanks!

File: app/controllers/images_controller.rb [hunk 2]
```diff
@@ -24,9 +24,9 @@ def fetch
   # assign attributes directly after rails update
   def create
     @warpable = Warpable.new
+    @warpable.history = 'None'
```
Comment (gauravano) [inline app/controllers/images_controller.rb]: What's the purpose of this attribute?

File: app/controllers/images_controller.rb [hunk 3]
```diff
@@ -72,28 +72,38 @@ def show
...
+        lat = point.split(',')[1]
+        node = Node.new(color: 'black',
+                        lat: lat,
+                        lon: lon,
+                        author: author,
+                        name: '')
+        node.save
+        nodes << node
+      end
+      @warpable.count_version = 0
```
Comment (gauravano) [inline app/controllers/images_controller.rb]: @divyabaid16 I gone through this PR and implementation seems neat. Great work, but I wasn't able to figure out the use of count_version. Please clarify.

**Comment by gauravano** [comment]
> Currently, when we click on a previous revision, it is cloned at put up on the top of stack. But this doesn't looks good.
So I will soon be making change to the code such that clicking on the version, it just goes back to that version without deleting all the forward versions.

Hmm, I think the current arrangement is fine. Suppose, a user want to see reverted version but after reverting they realize that previous configuration was better so, in that case, they can revert to that as last revision would be there. 

Because Paper::Trail.versions don't store the latest configuration, it's only present in the warpable, so on reverting it makes sense to store that last configuration from where we press revert. What do you all think?

File: app/controllers/images_controller.rb (hunk 2 above)
Comment (divyabaid16) [reply app/controllers/images_controller.rb]: This was just done to remove the MySQL2 error that says that ```history``` doesn't have any default value.

File: app/controllers/images_controller.rb (hunk 3 above)
Comment (divyabaid16) [reply app/controllers/images_controller.rb]: @gauravano 
So basically, this count_version is present in the update with a value zero. This is to make sure that the version is stored only when Updation is performed. Because currently, if I revert to a particular version, the X and the Y coordinates are changing and that is also considered as an update. So on reverting, one more cloned version is created. 

This was present just to specify that create a version whenever the value of ```version_count``` is zero. And for this to work, in ```revert``` function definition, I will have to add ```@warpable.count_version = 1``` so that it doesn't create any cloned copy and everything works in a smooth fashion.

So basically this is done just to prevent the extra cloning. Does it makes sense now?
Thanks!

**Comment by gauravano** [comment]
> currently, if I revert to a particular version, the X and the Y coordinates are changing and that is also considered as an update. So on reverting, one more cloned version is created.

Yes, the X and Y coordinates of warpable are changing as per the old version id passed. And, at the same time, a new version is created for the old configuration from where you triggered revert. I think, it's working fine as that last revision is a required and unique and not a clone.

**Comment by divyabaid16** [comment]
Okay, so now is it ready @gauravano ?

File: app/controllers/images_controller.rb [hunk 4]
```diff
@@ -72,28 +72,39 @@ def show
...
+  end
 
-    @warpable.nodes = nodes.collect(&:id).join(',')
-    @warpable.locked = params[:locked]
-    @warpable.cm_per_pixel = @warpable.get_cm_per_pixel
-    @warpable.save
-    render html: 'success'
+  def revert
+    @warpable = Warpable.find params[:id]
+    @warpable.count_version = 1
```
Comment (gauravano) [inline app/controllers/images_controller.rb]: Save the warpable to have effect of this

Comment (divyabaid16) [reply app/controllers/images_controller.rb]: Oops! sorry I forgot. I have made the changes @gauravano

**Comment by gauravano** [comment]
@divyabaid16 please write the expected behaviour for these actions:
- On update 
- On revert
- After reverting, updating the image
- After reverting, reverting again

**Comment by divyabaid16** [comment]
* On update: A new version of the image is created and it gets added to the list showing the versions.
* On revert: By clicking on which version to revert in the list, the image that will be displayed, will be of that version. 
* After reverting, updating the image: A new version of the image will be created.
* After reverting, reverting again: The last reverted version will be shown.

I think instead of using the term revert we can say it as *going back to that image without affecting the previous and future changes*.

**Comment by divyabaid16** [comment]
And for the follow up maybe we can add the option where the user can be provided the feature to lock a particular version, and all the changes made upto that version will be clubed to be a single version. And all the updates made after that will start with zero. What about this?

**Comment by gauravano** [comment]
Move `warpable.count_version` after reify as currently -- the count_version get changed to 1, we save the image and that warpable get stored in the versions table and it carries count_version = 1. And, warpable got reverted to old version and that version would be having count_version = 0`.

**Comment by gauravano** [comment]
What we want to achieve:
- If a user keeps on reverting in-between old version, no new version should be created at that time.
- Always create a version when update

**Comment by divyabaid16** [comment]
> What we want to achieve:
> 
>     * If a user keeps on reverting in-between old version, no new version should be created at that time.
> 
>     * Always create a version when update

Both of these conditions are followed.

**Comment by gauravano** [comment]
> If a user keeps on reverting in-between old version, no new version should be created at that time.

This condition isn't followed always like I tried reverting repeatedly and versions are getting created

Screenshot from 2019-07-28 20-04-44

File: app/models/warpable.rb [hunk 5]
```diff
@@ -18,6 +18,9 @@ class Warpable < ActiveRecord::Base
   belongs_to :map
   belongs_to :user
 
+  has_paper_trail on: %i(create update), only: %i(nodes),
+                  unless: proc { |t| t.count_version == 1 }
```
Comment (divyabaid16) [inline app/models/warpable.rb]: Due to this condition, version is not created when ```count_version==1``` 
It is created only when ```count_version==0``` which is only in the case of update. 
```suggestion
                  unless: proc { |t| t.count_version == 1 }
```
But if it's happening that's wierd.

File: app/controllers/images_controller.rb [hunk 6]
```diff
@@ -102,7 +102,7 @@ def update
   def revert
     @warpable = Warpable.find params[:id]
     version = @warpable.versions.find(params[:version])
-    version.reify&.save
+    version.reify
```
Comment (gauravano) [inline app/controllers/images_controller.rb]: This works good for all cases, except, we need to store a version when we first revert. So, do one thing -- Put if condition here something like `if count_version == 1`, then we are changing in-between reverts so do reify, `@warpable.count_version = 1` and save in this if block. Otherwise, reify.&save so that version would be created in case where we are reverting for first time. And, in update count_version = 0 is there so we would always create a version.

Comment (divyabaid16) [reply app/controllers/images_controller.rb]: On every update a new version is created and stored. And on revert, we are being taken to that version. So if I will use the if condition, a cloned copy of the reverted version would be added to the top of the stack, beacuse previously all the versions are already stored. I guess it would work fine without if. What do you think?

**Comment by jywarren** [comment]
> Currently, when we click on a previous revision, it is cloned at put up on the top of stack. But this doesn't looks good.
> So I will soon be making change to the code such that clicking on the version, it just goes back to that version without deleting all the forward versions.
> 
> So it would be like-- I click on the version, I'm reverted there. Then I click on the latest version, I'm reverted to the latest one. And it would work smoothly.

Thanks for thinking through this flow carefully @gauravano @divyabaid16!

Actually we use the "clone and put on top of stack" method in plots2's wiki system and it keeps things more linear, so I'm OK with that, esp. as it's already implemented. This may help avoid strange situations like "what happens to later revisions when you revert to an old revision?"

Like if we have A,B,C,D,E, and you revert to C, what do we do with D and E? In the clone scenario, we don't lose D and E, we just get F, which is a clone of C. That way, also, the reverting can be undone either by deleting F or just by cloning/reverting to E, AND it's clear who did the reversion since the clone bears the author id of the reverter, not the original author.

What do you think? does that simplify the discussion some? 

THANK YOU!!!

**Comment by divyabaid16** [comment]
This is exactly similar to what is implemented in Plots2. 
The only difference is that it is not cloned and put up on the top of the stack.

Here is how it works. 
I have version A B C D E F.
I revert to version C. Image of version C is displayed. Nothing is lost.
there is no change in the stack, it remains the same.
i.e A B C D E F, whether I revert to any version.

When I add a version, the stack becomes
A B C D E F F G, and now version G is displayed.

I revert to B, everything is present and B is displayed, and now if I make a new modification, it is added to the top of the stack.
It is also similar to how the Number of edits in a comment is displayed in Github. 

So, I guess it's ready if no further change is required.

Thanks!

**Comment by jywarren** [comment]
That's great - and I'm OK with it. However would the cloning system be
difficult to add? Because it has a distinct advantage of showing /who/
reverted as well as a timestamp for the revert. Thanks for the detailed
response!

**Comment by divyabaid16** [comment]
Ya, it won't be difficult to add the cloning system. So should I do it?

**Comment by jywarren** [comment]
Yes please! I think if it's not too hard let's just go for it. Thanks so
much!!!

**Comment by divyabaid16** [comment]
Hey @jywarren, it's done.
Thanks!

**Comment by jywarren** [comment]
Ok, are we good to merge here? Thanks!

File: app/controllers/images_controller.rb (hunk 6 above)
Comment (gauravano) [reply app/controllers/images_controller.rb]: When we revert for the first time, in that case without `reify.save`. That point from where we are reverting is getting lost that's why we want to save for that case.

File: app/controllers/images_controller.rb [hunk 7]
```diff
@@ -72,28 +72,44 @@ def show
...
 
-    @warpable.nodes = nodes.collect(&:id).join(',')
-    @warpable.locked = params[:locked]
-    @warpable.cm_per_pixel = @warpable.get_cm_per_pixel
+  def revert
+    @warpable = Warpable.find params[:id]
+    version = @warpable.versions.find(params[:version])
+    if(@warpable.count_version==1)
+      version.reify&.save
+    else
```
Comment (gauravano) [inline app/controllers/images_controller.rb]: Hey @divyabaid16, just read @jywarren's comments. If we want to save cloned versions then if condition is not required

Comment (gauravano) [reply app/controllers/images_controller.rb]: I guess,  count_version has no role to play now.

Comment (gauravano) [reply app/controllers/images_controller.rb]: Also, are you available for a quick call?

Comment (divyabaid16) [reply app/controllers/images_controller.rb]: Ya, Count_version has no role to play now. I will remove it.

Comment (gauravano) [reply app/controllers/images_controller.rb]: ok, so let's join on hangout?

Comment (divyabaid16) [reply app/controllers/images_controller.rb]: Okay @gauravano .

**Comment by jywarren** [comment]
Hi, just noting that we need to update the schema.rb file as that's what's used to build the db on production - without it, i got:

`ActionView::Template::Error (Mysql2::Error: Table 'mapknitter.versions' doesn't exist: SHOW FULL FIELDS FROM versions):`

Not a big deal, and we have a reminder for this on plots2 but not here (since we haven't been changing dbs that much!) but just noting for ourselves in the future.

Also, since I'm merging #912 now to fix a critical bug, this version tracking should be going live today !!!! 🎉
